libafl_bolts/src/rands.rs

```diff
@@ -7,14 +7,48 @@ use serde::{de::DeserializeOwned, Deserialize, Serialize};
 
 #[cfg(feature = "std")]
 use crate::current_nanos;
-#[cfg(any(feature = "xxh3", feature = "alloc"))]
-use crate::hash_std;
 
 /// The standard rand implementation for `LibAFL`.
 /// It is usually the right choice, with very good speed and a reasonable randomness.
 /// Not cryptographically secure (which is not what you want during fuzzing ;) )
 pub type StdRand = RomuDuoJrRand;
 
+/// Choose an item at random from the given iterator, sampling uniformly.
+///
+/// Note: the runtime cost is bound by the iterator's [`nth`][`Iterator::nth`] implementation
+///  * For `Vec`, slice, array, this is O(1)
+///  * For `HashMap`, `HashSet`, this is O(n)
+pub fn choose<I>(from: I, rand: u64) -> I::Item
+where
+    I: IntoIterator,
+    I::IntoIter: ExactSizeIterator,
+{
+    // create iterator
+    let mut iter = from.into_iter();
+
+    // make sure there is something to choose from
+    debug_assert!(iter.len() > 0, "choosing from an empty iterator");
+
+    // pick a random, valid index
+    let index = fast_bound(rand, iter.len() as u64) as usize;
+
+    // return the item chosen
+    iter.nth(index).unwrap()
+}
+
+/// Faster and almost unbiased alternative to `rand % n`.
+///
+/// For N-bit bound, probability of getting a biased value is 1/2^(64-N).
+/// At least 2^2*(64-N) samples are required to detect this amount of bias.
+///
+/// See: [An optimal algorithm for bounded random integers](https://github.com/apple/swift/pull/39143).
+#[must_use]
+pub fn fast_bound(rand: u64, n: u64) -> u64 {
+    debug_assert_ne!(n, 0);
+    let mul = u128::from(rand).wrapping_mul(u128::from(n));
```

Comment:
I don't know where I left the original comment, but: would u128 be an issue on 32 bit systems?
Also, this method should maybe be `#[inline]`(?)

Reply:
The compiler will inline it if it's good to do so, no worries.

Reply:
I'll check if `#[inline]` can help with cross-crate inlining without requiring LTO.

Regarding speed, how serious are we about 32-bit performance? Most of the PRNGs in the module already assume 64-bit system (and Lehmer64 even uses an 128x64 multiply).

What we can do is change the signature to `fast_bound(rand: u64, n: usize)`, and on 32-bit system fall back to full-width 32x32 multiply (as suggested by [Lemire](https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/)), instead of 64x64 like we do on 64-bit. This will give us the maximum possible speed on 32-bit systems, at the expense of more bias. I'd argue that this amount of bias is not a problem; usually `n` is much smaller than `2^32`.

Reply:
I like a 32 bit fallback since people sometimes run things on arm or even x86 32 bit code